eclair-core/src/test/scala/fr/acinq/eclair/channel/states/e/NormalSplicesStateSpec.scala

```diff
@@ -0,0 +1,188 @@
+      bob2alice.expectMsgType[TxSignatures]
+      bob2alice.forward(alice)
+    }
+
+    sender.expectMsgType[RES_SUCCESS[CMD_SPLICE]]
+    awaitCond(alice.stateData.asInstanceOf[DATA_NORMAL].spliceStatus == SpliceStatus.NoSplice)
+
+  }
+
+  test("recv CMD_SPLICE (splice-in)") { f =>
+    import f._
+
+    val initialState = alice.stateData.asInstanceOf[DATA_NORMAL]
+    assert(initialState.metaCommitments.main.capacity == 1_500_000.sat)
+    assert(initialState.metaCommitments.main.localCommit.spec.toLocal == 800_000_000.msat)
+    assert(initialState.metaCommitments.main.remoteCommit.spec.toLocal == 700_000_000.msat)
+
+    initiateSplice(f, spliceIn_opt = Some(SpliceIn(500_000 sat, pushAmount = 0 msat)))
+
+    assert(alice.stateData.asInstanceOf[DATA_NORMAL].metaCommitments.main.capacity == 2_000_000.sat)
+    assert(alice.stateData.asInstanceOf[DATA_NORMAL].metaCommitments.main.localCommit.spec.toLocal == 1_300_000_000.msat)
+    assert(alice.stateData.asInstanceOf[DATA_NORMAL].metaCommitments.main.remoteCommit.spec.toLocal == 700_000_000.msat)
+  }
+
+  test("recv CMD_SPLICE (splice-out)") { f =>
+    import f._
+
+    val initialState = alice.stateData.asInstanceOf[DATA_NORMAL]
+    assert(initialState.metaCommitments.main.capacity == 1_500_000.sat)
+    assert(initialState.metaCommitments.main.localCommit.spec.toLocal == 800_000_000.msat)
+    assert(initialState.metaCommitments.main.remoteCommit.spec.toLocal == 700_000_000.msat)
+
+    initiateSplice(f, spliceOut_opt = Some(SpliceOut(100_000 sat, ByteVector32.Zeroes)))
+
+    val feerate = TestConstants.Alice.nodeParams.onChainFeeConf.feeEstimator.getFeeratePerKw(TestConstants.Alice.nodeParams.onChainFeeConf.feeTargets.fundingBlockTarget)
+    val miningFee = Transactions.weight2fee(feerate, 600)
+    assert(alice.stateData.asInstanceOf[DATA_NORMAL].metaCommitments.main.capacity == 1_400_000.sat - miningFee)
+    assert(alice.stateData.asInstanceOf[DATA_NORMAL].metaCommitments.main.localCommit.spec.toLocal == 700_000_000.msat - miningFee)
+    assert(alice.stateData.asInstanceOf[DATA_NORMAL].metaCommitments.main.remoteCommit.spec.toLocal == 700_000_000.msat)
+  }
+
+  test("recv WatchFundingConfirmedTriggered with splice in progress", Tag(NoMaxHtlcValueInFlight)) { f =>
+    import f._
+
+    val sender = TestProbe()
+    // command for a large payment (larger than local balance pre-slice)
+    val cmd = CMD_ADD_HTLC(sender.ref, 1_000_000_000 msat, randomBytes32(), CltvExpiryDelta(144).toCltvExpiry(currentBlockHeight), TestConstants.emptyOnionPacket, None, localOrigin(sender.ref))
+    // first attempt at payment fails (not enough balance)
+    alice ! cmd
+    sender.expectMsgType[RES_ADD_FAILED[_]]
+    alice2bob.expectNoMessage()
+
+    initiateSplice(f, spliceIn_opt = Some(SpliceIn(500_000 sat, pushAmount = 0 msat)))
+    bob2alice.expectMsgType[Warning] // TODO: this is because non-initiator stops the InteractiveTx FSM right after having sent its signature
```

Comment:
> Btw, this leads to the re-sending of TxSignatures at restart. How do we manage that when we have multiple unconfirmed splices.

Why is that an issue? Can't we just look at the `txid` field of `tx_signatures` to know which splice it maps to?

> spawn an interactive-tx in the correct state on restart

I think this will be quite complex in practice. Simply adding a handler for `TxSignatures` in `NORMAL` whose only job is to transform a `PartiallySignedSharedTransaction` to a `FullySignedSharedTransaction` is much simpler IMO.

Reply:
> Why is that an issue? Can't we just look at the txid field of tx_signatures to know which splice it maps to?

I thought it would create a race between the finishing splice negotiation, and a potential new splice that is initiated immediately after, but, writing it down, I think that's fine. Yes what you are suggesting should work.

> Btw, this leads to the re-sending of `TxSignatures` at restart. How do we manage that when we have multiple unconfirmed splices?

Thinking about this a bit more. I initially thought we didn't have any way to tell, at restart, whether the channel was used after the last splice. But we instead should simply rely on local/remote locking status. As long as we haven't received remote's lock, we keep sending ours. WDYT?

Reply:
I think you're right, `splice_confirmed` is similar to `channel_ready`, so we should just send it when restarting and we should be fine?

Reply:
I'm talking about resending `TxSignatures` though, but I think it's the same. If the connection was down for a long time we would send `TxSignatures` and `SpliceConfirmed`.

Reply:
Handled `TxSignatures` for non-initiator in b99f2ee4fd3f75a66b6d407e1a9d417f3fd7eead.

> I think you're right, splice_confirmed is similar to channel_ready, so we should just send it when restarting and we should be fine?

I was wrong, unfortunately. It's not equivalent because a new commitment doesn't ack a splice like it does for a funding tx. Gave it a try in 6ac693f0575dbd87043dfa59261a5e230ca62b02 but there are flaws with this approach.

Reply:
I haven't revisited this yet, but after thinking a bit about it, I think just resending the most recent `SpliceLocked` at reconnection is fine. We just never ack them. A bit spammy maybe, but it's only a few dozen bytes anyway 🤷‍♂️.